Support TaskExt in CliApp

mihails-strasuns-sociomantic opened the pull request:
Misses release notes and tests but can already be tried out by early adopters ;)

joseph-wakeling-sociomantic commented on src/ocean/util/app/CliApp.d:
```diff
@@ -97,6 +107,24 @@ abstract class CliApp : Application, IArgumentsExtExtension
+        /***********************************************************************
+
+            Only used if `use_task_ext` is set to `true`. Defines scheduler
+            configuratio to be used by TaskExt.
```
Spelling: `configuration` ;-)

joseph-wakeling-sociomantic commented:
I'm trying this with a small CLI app and getting a runtime failure:

```
assertion failed at ./submodules/ocean/src/ocean/task/Scheduler.d:865
  toString():  Scheduler is null, initScheduler must be called before using it
Aborted (core dumped)
```

App constructor:

```D
    public this ()
    {
        typeof(this).OptionalSettings settings;
        settings.use_task_ext = true;
        super(this.app_name, this.app_desc, version_info, settings);
    }
```

App `run` method:

```D
    public override int run (Arguments args)
    {
        verify(Task.getThis() !is null);
        return 0;
    }
```

mihails-strasuns-sociomantic commented:
Updated:
- test
- release notes
- fixed implementation :)

joseph-wakeling-sociomantic commented:
Yea, this works with my app.  I'll re-review in detail once the couple of issues identified by CI are fixed.

mihails-strasuns-sociomantic commented:
Updated:
- fix 2.071 build

joseph-wakeling-sociomantic commented on integrationtest/taskext_cli/main.d:
```diff
@@ -0,0 +1,68 @@
+import ocean.text.Arguments;
+import ocean.task.Task;
+import ocean.task.Scheduler;
+import ocean.core.Test;
```
Minor style: imports can/should be in alphabetical order?

mihails-strasuns-sociomantic replied:
Nope, we don't require that. As a rule for automatic formatter that can be an option, but it is not worth bothering for manual effort.

joseph-wakeling-sociomantic commented on src/ocean/util/app/CliApp.d:
```diff
@@ -24,6 +24,7 @@ module ocean.util.app.CliApp;
 
 import ocean.util.app.Application : Application;
 import ocean.util.app.ext.model.IArgumentsExtExtension;
+import ocean.task.IScheduler;
```
Alphabetical order of imports?

joseph-wakeling-sociomantic commented on src/ocean/util/app/CliApp.d:
```diff
@@ -38,6 +39,7 @@ abstract class CliApp : Application, IArgumentsExtExtension
     import ocean.text.Arguments : Arguments;
     import ocean.util.app.ext.ArgumentsExt;
     import ocean.util.app.ext.VersionArgsExt;
+    import ocean.util.app.ext.TaskExt;
```
Alphabetical order of imports?

joseph-wakeling-sociomantic reviewed (approved):
Bar a couple of minor style issues, LGTM.

gavin-norman-sociomantic commented on src/ocean/util/app/ext/TaskExt.d:
```diff
@@ -44,6 +44,30 @@ import ocean.meta.codegen.Identifier;
+
+    ***************************************************************************/
+
+    SchedulerConfiguration config;
```
Why does `DaemonApp` not allow this to be set?

mihails-strasuns-sociomantic replied:
No particular reason - it just wasn't of much use when configuration file is present. I will add it for consistency.

mihails-strasuns-sociomantic replied:
Updated

gavin-norman-sociomantic replied:
:+1: yeah, that makes more sense. I wasn't sure whether there was some subtlety that I was missing.

mihails-strasuns-sociomantic committed:
Support scheduler settings in DaemonApp

gavin-norman-sociomantic commented on relnotes/taskextcli.feature.md:
````diff
@@ -25,3 +25,6 @@ class TestApp : CliApp
     }
 }
 ```
+
````
I'd put this as a separate note, otherwise `DaemonApp` users will miss it under the heading `CliApp now supports TaskExt`.

gavin-norman-sociomantic commented on src/ocean/util/app/ext/TaskExt.d:
```diff
@@ -44,6 +44,30 @@ import ocean.meta.codegen.Identifier;
 /// ditto
 class TaskExt : IConfigExtExtension
 {
+    /***************************************************************************
```
Maybe make this a separate commit?

mihails-strasuns-sociomantic committed:
Make possible to define scheduler defaults in TaskExt

Adds new argument to TaskExt constructor which will be used as
initial value for scheduler settings before config parsing takes
place.

mihails-strasuns-sociomantic committed:
Support TaskExt in CliApp

Fixes #532

mihails-strasuns-sociomantic commented:
Updated

gavin-norman-sociomantic commented:
Needs rebasing, apparently.

mihails-strasuns-sociomantic commented:
@gavin-norman-sociomantic I have suspicion github has updated UI to ask for rebasing even if there are no conflicts - only PRs that are based exactly on HEAD become green. You can still force rebase/merge it as administrator.

joseph-wakeling-sociomantic commented:
Yea, it's a safety mechanism that's intended to ensure CI is run for the PR features together with the very latest changes in the target branch.  FWIW we could probably guarantee the same result by making sure that we test merges into target branch as well as the PR alone (but of course GitHub can't tell if we are doing that, so it sounds this note of caution).
